SYS_ATL/new_eff.py

```diff
@@ -1262,6 +1273,290 @@ def bds(x,hi):
+#   /\ ( forall i,i'. May(InBound(i,i',e) /\ i < i')  =>
+#                     Commutes(a1', a2) /\ AllocCommutes(a1, a2) )
+#
+def Check_FissionLoop(proc, loop, stmts1, stmts2):
+    ctxt    = ContextExtraction(proc, [loop])
+
+    p       = ctxt.get_control_predicate()
+    G       = ctxt.get_pre_globenv()
+
+    slv     = SMTSolver(verbose=False)
+    slv.push()
+    slv.assume(AMay(p))
+
+    assert isinstance(loop, (LoopIR.ForAll,LoopIR.Seq))
+    i       = loop.iter
+    j       = i.copy()
+    hi      = loop.hi
+    subenv  = { i : LoopIR.Read(j, [], T.index, null_srcinfo()) }
+    stmts1_j= SubstArgs(stmts1, subenv).result()
+
+    a_bd    = expr_effs(hi)
+    a1      = stmts_effs(stmts1)
+    a1_j    = stmts_effs(stmts1_j)
+    a2      = stmts_effs(stmts2)
+
+    def bds(x,hi):
+        return AAnd(AInt(0) <= AInt(x), AInt(x) < lift_e(hi))
+
+    no_bound_change = (
+        AForAll([i], AImplies( AMay(bds(i,hi)),
+            AAnd(Commutes(a_bd, a1), Commutes(a_bd, a2))  )))
+    stmts_commute = (
+        AForAll([i,j], AImplies( AMay(AAnd( bds(i,hi), bds(j,hi),
+                                            AInt(i) < AInt(j) )),
+                                 AAnd(Commutes(a1_j, a2),
+                                      AllocCommutes(a1, a2)) )))
+
+    pred    = G(AAnd(no_bound_change, stmts_commute))
+    is_ok   = slv.verify(pred)
+    slv.pop()
+    if not is_ok:
+        raise SchedulingError(
+            f"Cannot fission loop over {i} at {loop.srcinfo}.")
+
+
+
+
+def Check_DeleteConfigWrite(proc, stmts):
+    assert len(stmts) > 0
+    ctxt = ContextExtraction(proc, stmts)
 
+    p       = ctxt.get_control_predicate()
+    G       = ctxt.get_pre_globenv()
+    ap      = ctxt.get_posteffs()
+    a       = G(stmts_effs(stmts))
+    stmtsG  = globenv(stmts)
+
+    slv     = SMTSolver(verbose=False)
+    slv.push()
+    a       = [E.Guard(AMay(p), a)]
+
+    # extract effects
+    WrG, Mod    = getsets([ES.WRITE_G, ES.MODIFY], a)
+    WrGp, RdGp  = getsets([ES.WRITE_G, ES.READ_G], ap)
+
+    # check that `stmts` does not modify any non-global data
+    only_mod_glob   = ADef(is_empty(LDiff(Mod,WrG)))
+    is_ok           = slv.verify(only_mod_glob)
+    if not is_ok:
+        slv.pop()
+        raise SchedulingError(
+            f"Cannot delete or insert statements at {stmts[0].srcinfo} "
+            f"because they may modify non-configuration data")
+
+    # get the set of config variables potentially modified by
+    # the statement block being focused on.  Filter out any
+    # such configuration variables whose values are definitely unchanged
+    def is_cfg_unmod_by_stmts(pt):
+        pt_e    = ABool(pt.name) if pt.typ == T.bool else AInt(pt.name) 
+        #cfg_unwritten = ADef( ANot(is_elem(pt, WrG)) )
+        cfg_unchanged = ADef( G(AEq(pt_e, stmtsG(pt_e))) )
+        return slv.verify(cfg_unchanged)
+    cfg_mod = set( pt.name for pt in get_point_exprs(WrG)
```

Comment:
This variable is never used, but it seems important...

Reply:
yup; copied down to the loop below it.  Thanks